cli/src/commands/extensions/api.rs

```diff
@@ -0,0 +1,257 @@
+//! TODO remove the following annotation before merging. The functions in
+//! this module should have as unique client the Deno runtime; pending its
+//! implementation, having these functions unused would break CI.
+#![allow(unused)]
+
+use std::cell::{RefCell, RefMut};
+use std::future::Future;
+use std::path::Path;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::commands::parse::{get_packages_from_lockfile, LOCKFILE_PARSERS};
+use crate::config::{get_current_project, Config};
+use crate::lockfiles::Parse;
+use crate::{api::PhylumApi, auth::UserInfo};
+
+use anyhow::{anyhow, Context, Error, Result};
+use deno_core::parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
+use deno_core::{op, OpDecl, OpState};
+use futures::future::BoxFuture;
+use once_cell::sync::Lazy;
+use phylum_types::types::auth::{AccessToken, RefreshToken};
+use phylum_types::types::common::{JobId, ProjectId};
+use phylum_types::types::job::JobStatusResponse;
+use phylum_types::types::package::{
+    Package, PackageDescriptor, PackageStatusExtended, PackageType,
+};
+use phylum_types::types::project::ProjectDetailsResponse;
+
+/// Holds either an unawaited, boxed `Future`, or the result of awaiting the future.
+/// In other words, it is a lazy evaluator for the `Future`.
+pub(crate) enum OnceFuture<T: Unpin> {
+    Future(BoxFuture<'static, T>),
+    Awaited(T),
+}
+
+impl<T: Unpin> OnceFuture<T> {
+    fn new(t: BoxFuture<'static, T>) -> Self {
+        OnceFuture::Future(t)
+    }
+
+    async fn get(&mut self) -> &T {
+        match *self {
+            OnceFuture::Future(ref mut t) => {
+                *self = OnceFuture::Awaited(t.await);
+                match *self {
+                    OnceFuture::Future(..) => unreachable!(),
+                    OnceFuture::Awaited(ref mut t) => t,
+                }
+            }
+            OnceFuture::Awaited(ref mut t) => t,
+        }
+    }
+}
+
+/// Opaquely encapsulates the extension state.
+pub struct ExtensionState(OnceFuture<Result<PhylumApi>>);
+
+impl ExtensionState {
+    /// Borrow an instance of `ExtensionState` from a reference to Deno's
+    /// `OpState` state handler, and retrieve a reference to the inner state.
+    /// The references lives as long as `&mut OpState`.
+    async fn borrow_from(state: &mut OpState) -> Result<&PhylumApi> {
+        state
+            .borrow_mut::<ExtensionState>()
+            .0
+            .get()
+            .await
+            .as_ref()
+            .map_err(|e| anyhow!("{:?}", e))
+    }
+}
```

Comment:
Having this method on `ExtensionState` seems **very** strange to me. There's basically nothing tying the `ExtensionState` struct to `OpState`, in fact `OpState` might not even have an `ExtensionState` stored.

Reply:
It might not be clear from this, but all ops do use `ExtensionState` as it is the object through which `PhylumApi` is accessed, so we need to provide this in `OpState` at runtime initialization. This specific function is called by most if not all ops, so not having it would create a lot of noise. We can look for alternative forms (`impl From<&mut OpState>` / `impl TryFrom<&mut OpState>` + `impl Deref<Target = PhylumApi>`?), but the substance of it is necessary.

Reply:
Actually, we may not use `impl From<T>` in this case as the associated function is asynchronous (that's probably the reason I used this form in the first place).

Reply:
The problem is just that `ExtensionState` here is neither input nor output, it's just a transitory maybe thing that should probably exist on `OpState`. There's various other ways to do this, either using traits or putting it on other structs, providing an unassociated method would work too. But putting it on `ExtensionState` doesn't feel right.

Reply:
I would very much like using known traits like `TryFrom` and `Deref`, but this is not possible as the crux of the issue is the await point in the middle of the chain. An unassociated function is really the only other viable option here but I don't really agree that that would provide semantic advantages over the current solution. I can try building a "smart pointer"/guard object to handle pinning like I mentioned in the other comment, which would probably be a better solution overall.